File: src/sprite_renderer.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const c = @import("c.zig");
const Shader = @import("shader.zig").Shader;
const Texture = @import("texture.zig").Texture;
const zm = @import("zmath");

pub const SpriteRenderer = struct {
    const Self = @This();
    shader: Shader,
    quad_vao: u32,

    pub fn init(allocator: *Allocator) Self {
        var shader = Shader.init();
        shader.create(allocator, .vertex, "resources/shaders/shader.vert");
        shader.create(allocator, .fragment, "resources/shaders/shader.frag");
        shader.createProgram();

        var vao: u32 = undefined;
        var vbo: u32 = undefined;
        var vertices = &[_]f32{
            // pos      // tex
            0.0, 1.0, 0.0, 1.0,
            1.0, 0.0, 1.0, 0.0,
            0.0, 0.0, 0.0, 0.0,

            0.0, 1.0, 0.0, 1.0,
            1.0, 1.0, 1.0, 1.0,
            1.0, 0.0, 1.0, 0.0,
        };

        c.glGenVertexArrays(1, &vao);
        c.glGenBuffers(1, &vbo);

        c.glBindBuffer(c.GL_ARRAY_BUFFER, vbo);
        c.glBufferData(c.GL_ARRAY_BUFFER, vertices.len * @sizeOf(f32), vertices, c.GL_STATIC_DRAW);

        c.glBindVertexArray(vao);
        c.glEnableVertexAttribArray(0);
        c.glVertexAttribPointer(0, 4, c.GL_FLOAT, c.GL_FALSE, 4 * @sizeOf(f32), null);
        c.glBindBuffer(c.GL_ARRAY_BUFFER, 0);
        c.glBindVertexArray(0);

        return Self{
            .quad_vao = vao,
            .shader = shader,
        };
    }

    pub fn deinit(self: *Self) void {
        c.glDeleteVertexArrays(self.quad_vao);
    }

    pub fn draw(self: *const Self, texture: u32, position: zm.Vec, scale: zm.Vec) void {
        self.shader.use();

        var model = zm.mul(zm.translationV(position), zm.scalingV(scale));
        self.shader.set_mat4("model", model);

        c.glActiveTexture(c.GL_TEXTURE0);
        Texture.bind(texture);

        c.glBindVertexArray(self.quad_vao);
        c.glDrawArrays(c.GL_TRIANGLES, 0, 6);
        c.glBindVertexArray(0);
    }
};

File: src/sprite_renderer.cpp
#include <glm/gtc/matrix_transform.hpp>
#include "sprite_renderer.h"
#include "texture.h"
#include "utils.h"

SpriteRenderer::SpriteRenderer(Shader *shader) {
	this->shader = shader;
	this->sprite_width = SPRITE_WIDTH;
	this->sprite_height = SPRITE_HEIGHT;
    unsigned int VBO;
    float vertices[] = {
        // pos      // tex
        0.0f, 1.0f, 0.0f, 1.0f,
        1.0f, 0.0f, 1.0f, 0.0f,
        0.0f, 0.0f, 0.0f, 0.0f,

        0.0f, 1.0f, 0.0f, 1.0f,
        1.0f, 1.0f, 1.0f, 1.0f,
        1.0f, 0.0f, 1.0f, 0.0f
    };

    glGenVertexArrays(1, &(this->quad_VAO));
    glGenBuffers(1, &VBO);

    glBindBuffer(GL_ARRAY_BUFFER, VBO);
    glBufferData(GL_ARRAY_BUFFER, sizeof(vertices), vertices, GL_STATIC_DRAW);

    glBindVertexArray(this->quad_VAO);
    glEnableVertexAttribArray(0);
    glVertexAttribPointer(0, 4, GL_FLOAT, GL_FALSE, 4 * sizeof(float), (void*)0);
    glBindBuffer(GL_ARRAY_BUFFER, 0);
    glBindVertexArray(0);
}

SpriteRenderer::~SpriteRenderer() {
    glDeleteVertexArrays(1, &(quad_VAO));
	delete shader;
}

void SpriteRenderer::render(Texture texture, glm::vec2 position, float rotation, glm::vec4 color, glm::vec2 scale) {
	shader->use();

	glm::mat4 model = glm::mat4(1.0f);
	model = glm::translate(model, glm::vec3(position[0], position[1], 0.0f));

    model = glm::translate(model, glm::vec3(0.5f * sprite_width, 0.5f * sprite_height, 0.0f));
    model = glm::rotate(model, glm::radians(rotation), glm::vec3(0.0f, 0.0f, 1.0f));
    model = glm::translate(model, glm::vec3(-0.5f * sprite_width, -0.5f * sprite_height, 0.0f));

	model = glm::scale(model, glm::vec3(scale.x, scale.y, 1.0f));

	shader->set_mat4((char *)"model", model);

    shader->set_vec4((char *)"spriteColor", color);

    glActiveTexture(GL_TEXTURE0);
    texture.bind();

    glBindVertexArray(quad_VAO);
    glDrawArrays(GL_TRIANGLES, 0, 6);
    glBindVertexArray(0);
}
